source-academy/cadet-frontend
# Standardise height behaviour of editor and REPL flexboxes

## UI/UX: Standardise height behaviour of editor and REPL flexboxes
Summary: Fixes #987 and an additional UI bug on both Firefox/Chrome.

### Description
The latest CSS working specification (see [here](https://www.w3.org/TR/css-flexbox-1/#min-size-auto)) states that `auto` is the default value for the `min-height` CSS property for flex items (in-flow elements of a flexbox). In the context of the flexboxes in the editor and REPL, `auto` is computed as the [*minimum content size*](https://www.w3.org/TR/css-sizing-3/#min-content), i.e. the minimum height such that the flex item does not exhibit vertical overflow.

The issue #987 thus describes the correct behaviour expected when the specification is followed (by Firefox), since the minimum height of the playground flex container grows as more REPL entries are added. However, this behaviour is not desirable here, as adding new entries to the REPL scrolls the page downwards and creates trailing dead space in the editor (whose height is clamped at 100%). This makes the editor-REPL combination clunky to use.

Chrome follows an outdated specification for flex items but exhibits the desired behaviour for users, whereby the REPL flex container is allowed to overflow and allow scrolling. However, it exhibits a different (but related) UI bug where a fixed amount of trailing dead space appears at the end of the editor when the REPL overflows (green box in screenshot below). This bug is also addressed by this PR.

![image](https://user-images.githubusercontent.com/44989315/82758712-bf0f2300-9e1a-11ea-9d13-7b05bf42b142.png)

To preserve the (incorrect but desired) behaviour exhibited by Chrome, we explicitly override the default value of `min-height` with `0` for specific ancestors of the editor and REPL flex containers. This is in line with the recommendation made by the Firefox issue [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1043520) and the Chrome issue [here](https://bugs.chromium.org/p/chromium/issues/detail?id=927066#c39). I have briefly checked the other pages in the Academy and it seems this change does not induce incorrect behaviour of other UI components.

### Changelog
- Fix nested flexboxes in the editor and REPL not respecting height of parent elements

### Verification instructions
Please test on **Firefox** (v70 or later) and **Chrome** (v75 or later) by:
1. Navigating to any page that has an editor and REPL, e.g. playground, assessment, submission grading
2. Create a long program in the editor that exceeds the editor height
3. Attempt to extend the REPL beyond its maximum height by executing commands, and verifying no dead space appears in the editor, and that the page does not scroll downward to accommodate new REPL entries

### Screenshots

#### Chrome (v81.0)
![image](https://user-images.githubusercontent.com/44989315/82758889-d4d11800-9e1b-11ea-96ad-c5f87d6bbcaa.png)

#### Firefox (v76.0.1)
![image](https://user-images.githubusercontent.com/44989315/82758869-b66b1c80-9e1b-11ea-85fd-5e4cb9576818.png)

Last updated 25 May 2020, 12:15AM
